core/server/data/migrations/versions/3.29/02-add-member-constraints.js

```diff
@@ -0,0 +1,103 @@
+const logging = require('../../../../../shared/logging');
+
+module.exports = {
+    config: {
+        transaction: true
+    },
+
+    async up({transacting: knex}) {
+        if (knex.client.config.client !== 'mysql') {
+            return logging.warn('Skipping member tables index creation - database is not MySQL');
+        }
+
+        const [membersLabelsIndexes] = await knex.raw('SHOW INDEXES FROM members_labels');
+        const [membersStripeCustomersIndexes] = await knex.raw('SHOW INDEXES FROM members_stripe_customers');
+        const [membersStripeCustomersSubscriptionsIndexes] = await knex.raw('SHOW INDEXES from members_stripe_customers_subscriptions');
+
+        if (membersLabelsIndexes.find(index => index.Key_name === 'members_labels_member_id_foreign')) {
+            logging.warn('Skipping "members_labels_member_id_foreign" foreign key constraint creation - already exists');
+        } else {
+            logging.info('Adding "members_labels_member_id_foreign" foreign key constraint');
+            await knex.schema.alterTable('members_labels', (table) => {
+                table.foreign('member_id').references('members.id').onDelete('CASCADE');
+            });
+        }
+
+        if (membersStripeCustomersIndexes.find(index => index.Key_name === 'members_stripe_customers_member_id_foreign')) {
+            logging.warn('Skipping "members_stripe_customers_member_id_foreign" foreign key constraint creation - already exists');
+        } else {
+            logging.info('Adding "members_stripe_customers_member_id_foreign" foreign key constraint');
+            await knex.schema.alterTable('members_stripe_customers', (table) => {
+                table.foreign('member_id').references('members.id').onDelete('CASCADE');
+            });
+        }
+
+        if (membersStripeCustomersIndexes.find(index => index.Key_name === 'members_stripe_customers_customer_id_unique')) {
+            logging.warn('Skipping "members_stripe_customers_customer_id_unique" index creation - already exists');
+        } else {
+            logging.info('Adding "members_stripe_customers_customer_id_unique" index');
+            await knex.schema.alterTable('members_stripe_customers', (table) => {
+                table.unique('customer_id');
+            });
+        }
+
+        if (membersStripeCustomersSubscriptionsIndexes.find(index => index.Key_name === 'members_stripe_customers_subscriptions_customer_id_foreign')) {
+            logging.warn('Skipping "members_stripe_customers_subscriptions_customer_id_foreign" foreign key constraint creation - already exists');
+        } else {
+            logging.info('Adding "members_stripe_customers_subscriptions_customer_id_foreign" foreign key constraint');
+            await knex.schema.alterTable('members_stripe_customers_subscriptions', (table) => {
+                table.foreign('customer_id').references('members_stripe_customers.customer_id').onDelete('CASCADE');
```

Comment:
Did we do any checks on real-world data if there are any "orphaned" `members_stripe_customers_subscriptions` records? I don't think this situation is very likely, but had a state on my local db where I got one record in `members_stripe_customers_subscriptions` table without corresponding `members_stripe_customer`.

Reply:
Not that I'm aware of unless @allouis did any during the de-duplication work. Do you think it's necessary to have another migration to remove any orphans?

Reply:
Really good point, it _is_ currently possible to have orphaned rows if you deleted members with subscriptions pre Ghost 3.27. 

@kevinansfield we'll need a migration to clean up orphaned rows, but maybe that needs more discussion?

Reply:
Renamed to `05-` to prepare for orphaned row deletion migrations